extensions/elytron-security-jdbc/runtime/src/main/java/io/quarkus/elytron/security/jdbc/JdbcRecorder.java

```diff
@@ -63,8 +61,9 @@ private void registerPrincipalQuery(PrincipalQueryConfig principalQuery, JdbcSec
     }
 
     private DataSource getDataSource(PrincipalQueryConfig principalQuery) {
-        return (DataSource) principalQuery.datasource
-                .map(name -> Arc.container().instance(name).get())
+        return principalQuery.datasource
+                .map(name -> Arc.container()
+                        .instance(DataSource.class, new io.quarkus.agroal.DataSource.DataSourceLiteral(name)).get())
```

Comment:
Sorry a bit late for that comment but I'm wondering if we should get rid of this lambda while we're at it. It really doesn't bring anything useful. I can do it later on if you're tired of my nitpicking :).

Reply:
Oh boy, you are right! I didn't see this was runtime code, let's fix it.

Reply:
Fixed